crates/bevy_reflect/bevy_reflect_derive/src/impls/enums.rs

```diff
@@ -275,40 +275,53 @@ fn generate_impls(reflect_enum: &ReflectEnum, ref_index: &Ident, ref_name: &Iden
         let name = ident.to_string();
         let unit = reflect_enum.get_unit(ident);
 
-        // This is equivalent to a |fields: &Vec<StructField>, to_run: |usize, usize, &StructField| -> TokenStream|
-        // closure. Issue is that the closure cannot itself accept another closure, because closures cannot vary in
-        // the concrete type they accept, and each closure has a different concrete type.
-        macro_rules! for_fields {
-            ($fields:expr, |$reflect_idx:ident, $declar_field:pat, $field:ident| $body:expr) => {{
-                let mut field_info = Vec::new();
-                let mut $reflect_idx: usize = 0;
-                for ($declar_field, $field) in $fields.iter().enumerate() {
-                    if $field.attrs.ignore {
-                        // Ignored field
-                        continue;
-                    }
-                    field_info.push($body);
-                    $reflect_idx += 1;
+        fn for_fields(
+            fields: &[StructField],
+            mut generate_for_field: impl FnMut(usize, usize, &StructField) -> proc_macro2::TokenStream,
+        ) -> (usize, Vec<proc_macro2::TokenStream>) {
+            let mut constructor_argument = Vec::new();
+            let mut reflect_idx = 0;
+            for field in fields.iter() {
+                if field.attrs.ignore {
+                    // Ignored field
+                    continue;
                 }
-                ($reflect_idx, field_info)
-            }};
+                constructor_argument.push(generate_for_field(reflect_idx, field.index, field));
+                reflect_idx += 1;
+            }
+            (reflect_idx, constructor_argument)
         }
-        let (variant_type, field_len, field_info) = match &variant.fields {
-            EnumVariantFields::Unit => ("Unit", 0usize, quote!(#name)),
+        let mut info_type = |variant, info_type, arguments| {
+            let variant = Ident::new(variant, Span::call_site());
+            let info_type = Ident::new(info_type, Span::call_site());
+            variant_info.push(quote! {
+                #bevy_reflect_path::VariantInfo::#variant(
+                    #bevy_reflect_path::#info_type::new(#arguments)
+                )
+            });
+            variant
+        };
```

Comment:
Nice, yeah I think this makes each arm more readable. Thanks!

I know this is becoming pointless bikeshedding (my apologies haha), but do you think it makes sense to put everything (e.g. `enum_variant_type.push(...)`, etc.) else in this closure? Or perhaps that just puts us back to where we started and I should just try to not be so finicky with stuff like this 😅

Reply:
Oh I see. Great idea.